components/script/dom/visualviewport.rs

```diff
@@ -0,0 +1,160 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::Cell;
+
+use dom_struct::dom_struct;
+use euclid::Rect;
+use script_bindings::codegen::GenericBindings::VisualViewportBinding::VisualViewportMethods;
+use script_bindings::num::Finite;
+use script_bindings::root::{Dom, DomRoot};
+use script_bindings::script_runtime::CanGc;
+use style_traits::CSSPixel;
+
+use crate::dom::bindings::reflector::reflect_dom_object;
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::window::Window;
+
+/// <https://drafts.csswg.org/cssom-view/#the-visualviewport-interface>
+#[dom_struct]
+pub(crate) struct VisualViewport {
+    eventtarget: EventTarget,
+
+    /// The associated [`Window`] for this visual viewport.
+    window: Dom<Window>,
+
+    /// The rectangle bound of the visual viewport, relative to the layout viewport.
+    #[no_trace]
+    viewport_rect: Cell<Rect<f32, CSSPixel>>,
+
+    /// The scale factor of visual viewport, which is also commonly known as pinch-zoom.
+    /// <https://drafts.csswg.org/cssom-view/#scale-factor>
+    scale: Cell<f32>,
```

Comment:
Now that I think about this (sorry for not noticing it before), isn't scale essentially just the size of one dimension viewport_rect divided by the same dimension in the `Window` viewport? Maybe you don't even need to store this.